README.md
[![Build Status](https://secure.travis-ci.org/STRML/node-toobusy.png)](http://travis-ci.org/STRML/node-toobusy)

# Is Your Node Process Too Busy?

`toobusy-js` is a fork of lloyd's [node-toobusy](http://github.com/lloyd/node-toobusy) that removes native dependencies
in favor of using the `unref` introduced in [node 0.9.1](http://blog.nodejs.org/2012/08/28/node-v0-9-1-unstable/).

This package is a simpler install without native dependencies, but requires node >= 0.9.1.

## Node-Toobusy

What happens when your service is overwhelmed with traffic?
Your server can do one of two things:

  * Stop working, or...
  * Keep serving as many requests as possible

This library helps you do the latter.

## How it works

`toobusy` polls the node.js event loop and keeps track of "lag",
which is how long requests wait in node's event queue to be processed.
When lag crosses a threshold, `toobusy` tells you that you're *too busy*.
At this point you can stop request processing early
(before you spend too much time on them and compound the problem),
and return a "Server Too Busy" response.
This allows your server to stay *responsive* under extreme load,
and continue serving as many requests as possible.

## installation

```
npm install toobusy-js
```


## usage

```javascript
var toobusy = require('toobusy-js'),
    express = require('express');

var app = express();

// middleware which blocks requests when we're too busy
app.use(function(req, res, next) {
  if (toobusy()) {
    res.send(503, "I'm busy right now, sorry.");
  } else {
    next();
  }
});

app.get('/', function(req, res) {
  // processing the request requires some work!
  var i = 0;
  while (i < 1e5) i++;
  res.send("I counted to " + i);
});

var server = app.listen(3000);

process.on('SIGINT', function() {
  server.close();
  // calling .shutdown allows your process to exit normally
  toobusy.shutdown();
  process.exit();
});
```

## tunable parameters

The library exposes a few knobs:

`maxLag` - This number represents the maximum amount of time in milliseconds that the event queue is behind,
before we consider the process *too busy*.
`interval` - The check interval for measuring event loop lag, in ms.

```javascript
var toobusy = require('toobusy-js');

// Set maximum lag to an aggressive value.
toobusy.maxLag(10);

// Set check interval to a faster value. This will catch more latency spikes
// but may cause the check to be too sensitive.
toobusy.interval(250);

// Get current maxLag or interval setting by calling without parameters.
var currentMaxLag = toobusy.maxLag(), interval = toobusy.interval();

toobusy.onLag(function(currentLag) {
  console.log("Event loop lag detected! Latency: " + currentLag + "ms");
});
```

The default maxLag value is 70ms, and the default check interval is 500ms.
This allows an "average" server to run at 90-100% CPU
and keeps request latency at around 200ms.
For comparison, a maxLag value of 10ms results in 60-70% CPU usage,
while latency for "average" requests stays at about 40ms.

These numbers are only examples,
and the specifics of your hardware and application can change them drastically,
so experiment!
The default of 70 should get you started.

## Events

As of `0.5.0`, `toobusy-js` exposes an `onLag` method. Pass it a callback to be notified when
a slow event loop tick has been detected.

## references

> There is nothing new under the sun. (Ecclesiastes 1:9)

Though applying "event loop latency" to node.js was not directly inspired by anyone else's work,
this concept is not new.  Here are references to others who apply the same technique:

  * [Provos, Lever, and Tweedie 2000](http://www.kegel.com/c10k.html#tips) - "notes that dropping incoming connections when the server is overloaded improved the shape of the performance curve."

## license

[WTFPL](http://wtfpl.org)


CHANGELOG.md
### 0.6.3
  * Update example.js. Thanks @khalidsalomao

### 0.6.2
  * New example and improvements in package.json. Thanks @khalidsalomao

### 0.6.1
  * New smoothing factor to be used when lag is falling
  * Isolate logic of currentLag in an overwritable function

### 0.5.1
  * Set `onLag` default threshould to `maxLag()`. Thanks @flentini

### 0.5.0
  * Add `toobusy.onLag()` for catching lag events. Thanks @mnahkies

### 0.4.3
  * Update dampening formula and make smoothing tunable. Thanks @tgroleau

### 0.4.2
  * Input validation for `maxLag()` and `interval()`.

### 0.4.1
  * Fix checking not automatically starting.

### 0.4.0
  * Refactoring
  * Added `toobusy.interval()` for getting/setting check interval.

### 0.3.0
  * full JS implementation, removed c code & bindings
  * example bugfixes

### 0.2.3
  * node.js 0.10.0 support upgrade to latest bindings and include in .travis.yml

### 0.2.2
  * *really* works on windows
  * improved documentation

### 0.2.1
  * works on windows
  * improved express.js example (now targets express 3.0)

### 0.2.0
  * documentation improvements
  * improved examples (including load generation tool and server)
  * .lag() added to retrieve (periodically calculated) event loop lag in ms

### 0.1.1
  * hey look, unit tests!

### 0.1.0
  * improved documentation and samples
  * added `maxLag` parameter


tests.js
'use strict';
var should = require('should');
var toobusy = require('./');

function tightWork(duration) {
  var start = Date.now();
  while ((Date.now() - start) < duration) {
    for (var i = 0; i < 1e5;) i++;
  }
}

/*global describe, it, beforeEach, afterEach */
describe('the library', function() {
  it('should export a couple functions', function() {
    (toobusy).should.be.a('function');
    (toobusy.maxLag).should.be.a('function');
    (toobusy.shutdown).should.be.a('function');
    (toobusy.interval).should.be.a('function');
    (toobusy.shutdown).should.be.a('function');
    (toobusy).should.not.have.property('start');
  });
  it('should start automatically', function() {
    (toobusy.started()).should.equal(true);
  });
});

describe('maxLag', function() {
  it('should default to 70', function() {
    (toobusy.maxLag()).should.equal(70);
  });
  it('should throw an exception for non-numbers', function() {
    (function() { toobusy.maxLag('derp'); }).should.throw(/must be a number/);
  });
  it('should throw an exception for values < 10', function() {
    (function() { toobusy.maxLag(9); }).should.throw(/should be greater than 10/);
  });
  it('should be configurable', function() {
    (toobusy.maxLag(50)).should.equal(50);
    (toobusy.maxLag(10)).should.equal(10);
    (toobusy.maxLag()).should.equal(10);
  });
});

describe('interval', function() {
  it('should default to 500', function() {
    (toobusy.interval()).should.equal(500);
  });
  it('should throw an exception for values < 16', function() {
    (function() { toobusy.interval(15); }).should.throw(/Interval/);
  });
  it('should be configurable', function() {
    (toobusy.interval(250)).should.equal(250);
    (toobusy.interval(300)).should.equal(300);
    (toobusy.interval()).should.equal(300);
  });
});

describe('toobusy()', function() {
  // Set lower thresholds for each of these tests.
  // Resetting the interval() also resets the internal lag counter, which
  // is nice for making these tests independent of each other.
  beforeEach(function() {
    toobusy.maxLag(10);
    toobusy.interval(250);
  });
  afterEach(function() {
    toobusy.maxLag(70);
    toobusy.interval(500);
  });
  it('should return true after a little load', function(done) {
    function load() {
      if (toobusy()) return done();
      tightWork(100);
      setTimeout(load, 0);
    }
    load();
  });

  it('should return a lag value after a little load', function(done) {
    function load() {
      if (toobusy()) {
        var lag = toobusy.lag();
        should.exist(lag);
        lag.should.be.above(1);
        return done();
      }
      tightWork(100);
      setTimeout(load, 0);
    }
    load();
  });

  describe('lag events', function () {
    //Sometimes the default 2s timeout is hit on this suite, raise to 10s.
    this.timeout(10 * 1000);

    it('should not emit lag events if the lag is less than the configured threshold',
        testLagEvent(100, 50, false));
    it('should emit lag events if the lag is greater than the configured threshold',
        testLagEvent(50, 100, true));
    it('should emit lag events if lag occurs and no threshold is specified',
        testLagEvent(undefined, 500, true));

    function testLagEvent(threshold, work, expectFire) {
      return function (done) {
        var calledDone = false;

        toobusy.onLag(function (lag) {
          if (calledDone) {
            return;
          }

          if (!expectFire) {
            calledDone = true;
            done(new Error('lag event fired unexpectedly'));
            return;
          }

          should.exist(lag);
          lag.should.be.above(threshold || 0);

          calledDone = true;
          done();
        }, threshold);

        if (!expectFire) {
          setTimeout(function () {
            if (!calledDone) {
              calledDone = true;
              done();
            }
          }, work + threshold);
        }

        tightWork(work);
      }
    }
  });
});

describe('smoothingFactor', function() {
  //Sometimes the default 2s timeout is hit on this suite, raise to 10s.
  this.timeout(10 * 1000);

  beforeEach(function() {
    toobusy.reset();
    toobusy.maxLag(10);
    toobusy.interval(250);
  });
  afterEach(function() {
    toobusy.maxLag(70);
    toobusy.interval(500);
  });

  setupSmoothingFactorTests({
    function: toobusy.smoothingFactorOnRise,
    suiteName: 'on rise',
    default: 1/3
  });

  setupSmoothingFactorTests({
    function: toobusy.smoothingFactorOnFall,
    suiteName: 'on fall',
    default: 1 - 1/3
  });

  it('should allow no dampening', function (done) {
    var cycles_to_toobusy = 0;
    toobusy.smoothingFactorOnRise(1); // no dampening
    toobusy.smoothingFactorOnFall(1); // no dampening

    function load() {
      if (toobusy()) {
        (cycles_to_toobusy).should.equal(3);
        return done();
      }
      cycles_to_toobusy++;
      tightWork(100); // in 3 ticks, will overshoot by ~50ms, above 2*10ms
      setImmediate(load);
    }

    load();
  });
  it('should respect larger dampening factors', function (done) {
    var cycles_to_toobusy = 0;
    toobusy.smoothingFactorOnRise(0.05);

    function load() {
      if (toobusy()) {
        (cycles_to_toobusy).should.be.above(3);
        return done();
      }
      cycles_to_toobusy++;
      tightWork(100);
      setImmediate(load);
    }

    load();
  });

});

function setupSmoothingFactorTests(options) {
  var smoothingFunc = options.function;

  describe(options.suiteName, function() {
    it('should default to ' + options.default, function(done) {
      (smoothingFunc()).should.equal(options.default);
      done();
    });
    it('should throw an exception for invalid values', function(done) {
      (function() { smoothingFunc(0); }).should.throw;
      (function() { smoothingFunc(2); }).should.throw;
      (function() { smoothingFunc(-1); }).should.throw;
      (function() { smoothingFunc(1); }).should.not.throw;
      done();
    });
    it('should be configurable', function(done) {
      (smoothingFunc(0.9)).should.equal(0.9);
      (smoothingFunc(0.1)).should.equal(0.1);
      (smoothingFunc()).should.equal(0.1);
      done();
    });
  });
};

describe('started', function() {
  it('should return false after shutdown', function(done) {
    toobusy.shutdown();
    (toobusy.started()).should.equal(false);
    done();
  });
});




toobusy.js
'use strict';

var events = require('events');

//
// Constants
//
var STANDARD_HIGHWATER = 70;
var STANDARD_INTERVAL = 500;
var LAG_EVENT = "LAG_EVENT";

// A dampening factor.  When determining average calls per second or
// current lag, we weigh the current value against the previous value 2:1
// to smooth spikes.
// See https://en.wikipedia.org/wiki/Exponential_smoothing
var SMOOTHING_FACTOR = 1/3;

//
// Vars
//

var lastTime = Date.now();
var highWater = STANDARD_HIGHWATER;
var interval = STANDARD_INTERVAL;
var smoothingFactorOnRise = SMOOTHING_FACTOR;
var smoothingFactorOnFall = 1 - SMOOTHING_FACTOR;
var currentLag = 0;
var checkInterval;
var lagEventThreshold = -1;
var eventEmitter = new events.EventEmitter();

/**
 * Main export function.
 * @return {Boolean} True if node process is too busy.
 */
var toobusy = function(){
  // If current lag is < 2x the highwater mark, we don't always call it 'too busy'. E.g. with a 50ms lag
  // and a 40ms highWater (1.25x highWater), 25% of the time we will block. With 80ms lag and a 40ms highWater,
  // we will always block.
  var pctToBlock = (currentLag - highWater) / highWater;
  return Math.random() < pctToBlock;
};

/**
 * Resets variables to their default values
 * Affected variables: highWater, interval, smoothingFactorOnRise and smoothingFactorOnFall
 */
toobusy.reset = function() {
  highWater = STANDARD_HIGHWATER;
  interval = STANDARD_INTERVAL;
  smoothingFactorOnRise = SMOOTHING_FACTOR;
  smoothingFactorOnFall = 1 - SMOOTHING_FACTOR;
};

/**
 * Sets or gets the current check interval.
 * If you want more sensitive checking, set a faster (lower) interval. A lower maxLag can also create a more
 * sensitive check.
 * @param  {Number} [newInterval] New interval to set. If not provided, will return the existing interval.
 * @return {Number}               New or existing interval.
 */
toobusy.interval = function(newInterval) {
  if (!newInterval) return interval;
  if (typeof newInterval !== "number") throw new Error("Interval must be a number.");

  newInterval = Math.round(newInterval);
  if(newInterval < 16) throw new Error("Interval should be greater than 16ms.");

  toobusy.shutdown();
  interval = newInterval;
  start();
  return interval;
};

/**
 * Returns last lag reading from last check interval.
 * @return {Number} Lag in ms.
 */
toobusy.lag = function(){
  return Math.round(currentLag);
};

/**
 * Set or get the current max latency threshold. Default is 70ms.
 *
 * Note that if event loop lag goes over this threshold, the process is not always 'too busy' - the farther
 * it goes over the threshold, the more likely the process will be considered too busy.
 *
 * The percentage is equal to the percent over the max lag threshold. So 1.25x over the maxLag will indicate
 * too busy 25% of the time. 2x over the maxLag threshold will indicate too busy 100% of the time.
 * @param  {Number} [newLag] New maxLag (highwater) threshold.
 * @return {Number}          New or existing maxLag (highwater) threshold.
 */
toobusy.maxLag = function(newLag){
  if(!newLag) return highWater;

  // If an arg was passed, try to set highWater.
  if (typeof newLag !== "number") throw new Error("MaxLag must be a number.");
  newLag = Math.round(newLag);
  if(newLag < 10) throw new Error("Maximum lag should be greater than 10ms.");

  highWater = newLag;
  return highWater;
};

/**
 *  Private function used to set the two smoothing factors
 * @param {number} newFactor
 * @param {boolean} onFallSelector Optional parameter used to selected which factor to set
 */
function setSmoothingFactor(newFactor, onFallSelector){
  if (typeof newFactor !== "number") throw new Error("NewFactor must be a number.");
  if(newFactor <= 0 || newFactor > 1) throw new Error("Smoothing factor should be in range ]0,1].");

  if (onFallSelector) {
    smoothingFactorOnFall = newFactor;
  } else {
    smoothingFactorOnRise = newFactor;
  }
};

/**
 * Set or get the smoothing factor on rise. Default is 0.3333....
 * This is the factor applied when the measured lag is higher than currentLag
 *
 * The smoothing factor per the standard exponential smoothing formula "αtn + (1-α)tn-1"
 * See: https://en.wikipedia.org/wiki/Exponential_smoothing
 *
 * @param  {Number} [newFactor] New smoothing factor.
 * @return {Number}             New or existing smoothing factor.
 */
toobusy.smoothingFactorOnRise = function(newFactor) {
  if(!newFactor) return smoothingFactorOnRise;
  setSmoothingFactor(newFactor);
  return smoothingFactorOnRise;
};

/**
 * Set or get the smoothing factor on fall. Default is 0.6666....
 * This is the factor applied when the measured lag is lower than currentLag
 *
 * @param  {Number} [newFactor] New smoothing factor.
 * @return {Number}             New or existing smoothing factor.
 */
toobusy.smoothingFactorOnFall = function(newFactor) {
  if(!newFactor) return smoothingFactorOnFall;
  setSmoothingFactor(newFactor, true);
  return smoothingFactorOnFall;
};

/**
 * Shuts down toobusy.
 *
 * Not necessary to call this manually, only do this if you know what you're doing. `unref()` is called
 * on toobusy's check interval, so it will never keep the server open.
 */
toobusy.shutdown = function(){
  currentLag = 0;
  checkInterval = clearInterval(checkInterval);
};

toobusy.started = function() {
  return !!checkInterval;
};

/**
 * Registers an event listener for lag events,
 * optionally specify a minimum value threshold for events being emitted
 * @param {Function}  fn  Function of form onLag(value: number) => void
 * @param {number}  [threshold=maxLag] Optional minimum lag value for events to be emitted
 */
toobusy.onLag = function (fn, threshold) {

  if (typeof threshold === "number") {
    lagEventThreshold = threshold;
  } else {
    lagEventThreshold = toobusy.maxLag();
  }

  eventEmitter.on(LAG_EVENT, fn);
};

/**
 * Calculates the new value that will be assigned to currentLag
 * Overwrite this function if you need a different behavior
 * @param {number} lag The last measured lag
 * @param {number} cLag The currentLag value
 * @param {number} sFactor The smoothingFactor value
 * @return {number} The calculated value
 */
toobusy.lagFunction = function (lag, cLag, sFactorRise, sFactorFall) {
    var factor = lag > cLag ? sFactorRise : sFactorFall;
    return factor * lag + (1 - factor) * cLag;
}

/**
 * Private - starts checking lag.
 */
function start() {
  checkInterval = setInterval(function(){
    var now = Date.now();
    var lag = now - lastTime;
    lag = Math.max(0, lag - interval);
    currentLag = toobusy.lagFunction(lag, currentLag, smoothingFactorOnRise, smoothingFactorOnFall);
    lastTime = now;

    if (lagEventThreshold !== -1 && currentLag > lagEventThreshold) {
      eventEmitter.emit(LAG_EVENT, currentLag);
    }

  }, interval);

  // Don't keep process open just for this timer.
  checkInterval.unref();
}

// Kickoff the checking!
start();

module.exports = toobusy;
